src/providers/coinbase.rs

```diff
@@ -0,0 +1,137 @@
+use {
+    super::HistoryProvider,
+    crate::{
+        error::{RpcError, RpcResult},
+        handlers::{
+            HistoryQueryParams,
+            HistoryResponseBody,
+            HistoryTransaction,
+            HistoryTransactionMetadata,
+        },
+    },
+    async_trait::async_trait,
+    axum::{body::Bytes, http::method},
+    ethers::types::H160,
+    futures_util::StreamExt,
+    hyper::Client,
+    hyper_tls::HttpsConnector,
+    serde::{Deserialize, Serialize},
+    tracing::{info, log::error},
+    url::Url,
+};
+
+#[derive(Debug)]
+pub struct CoinbaseProvider {
+    pub api_key: String,
+    pub app_id: String,
+    pub http_client: Client<HttpsConnector<hyper::client::HttpConnector>>,
+}
+
+impl CoinbaseProvider {
+    pub fn new(api_key: String, app_id: String) -> Self {
+        let http_client = Client::builder().build::<_, hyper::Body>(HttpsConnector::new());
+        Self {
+            api_key,
+            app_id,
+            http_client,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct CoinbaseResponseBody {
+    pub transactions: Vec<CoinbaseTransaction>,
+    pub next_page_key: Option<String>,
+    pub total_count: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct CoinbaseTransaction {
+    pub status: String,
+    pub transaction_id: String,
+    pub tx_hash: String,
+    pub created_at: String,
+}
+
+#[async_trait]
+impl HistoryProvider for CoinbaseProvider {
+    async fn get_transactions(
+        &self,
+        address: H160,
+        body: Bytes,
+        params: HistoryQueryParams,
+    ) -> RpcResult<HistoryResponseBody> {
+        let base = format!(
+            "https://pay.coinbase.com/api/v1/buy/user/{}/transactions",
+            &address.to_string()
+        );
```

Comment:
Pretty sure this could be simplified?

```suggestion
        let base = format!("https://pay.coinbase.com/api/v1/buy/user/{address}/transactions");
```